### PR TITLE
Fixed Integrated Eye Sensors On Examining

### PR DESCRIPTION
**Pull request by Sparkyhotdog**
This fixes eye sensors not showing the appropriate records on examining someone.
Edit: Now does all HUDs, because there were a few of them

### IC changelog

- bugfix: "All medical and security HUDs now display records on examining."

### CODE REVIEW

**Comment by Sparkyhotdog** [comment]
!review

**Comment by Snakebittenn** [comment]
You should extend this to all instances of having a HUD.

**Comment by Sparkyhotdog** [comment]
> You should extend this to all instances of having a HUD.

Could you explain? I thought I had done this...

**Comment by Snakebittenn** [comment]
For instance, aviators/hudpatches.

**Comment by Sparkyhotdog** [comment]
They already do.

**Comment by Alberyk** [comment]
This is a not a very good solution. Adding those vars to the mob is not really the best way to go. I would suggest a check to see if the mob can edit records, in which the proc looks for the glasses and the implant/organ with the right choice.

**Comment by Sparkyhotdog** [comment]
!bugfix

File: code/modules/mob/living/carbon/human/examine.dm [hunk 1]
```diff
@@ -364,11 +364,16 @@
 /proc/hasHUD(mob/M as mob, hudtype)
 	if(istype(M, /mob/living/carbon/human))
 		var/mob/living/carbon/human/H = M
+		var/eye_hud
+		var/obj/item/clothing/glasses/G = H.glasses
+		// Checks for eye sensor HUD
+		for (var/obj/item/organ/internal/augment/eye_sensors/organ in H.internal_organs)
+			eye_hud = organ.selected_hud
```
Comment (Geevies) [inline code/modules/mob/living/carbon/human/examine.dm:371]:
```suggestion
		var/obj/item/organ/internal/augment/eye_sensors/E = locate() in H.internal_organs
		if(E)
			eye_hud = E.selected_hud
```

**Review by Geevies — APPROVED** [review]
make sure to test it again, but it looks good

**Comment by Sparkyhotdog** [comment]
Tested. All working.

File: code/modules/mob/living/carbon/human/examine.dm [hunk 2]
```diff
@@ -364,11 +364,17 @@
...
+		var/eye_hud
+		var/obj/item/clothing/glasses/G = H.glasses
+		// Checks for eye sensor HUD
+		var/obj/item/organ/internal/augment/eye_sensors/E = locate() in H.internal_organs
+		if(E)
+			eye_hud = E.selected_hud
 		switch(hudtype)
 			if("security")
-				return istype(H.glasses, /obj/item/clothing/glasses/hud/security) || istype(H.glasses, /obj/item/clothing/glasses/sunglasses/sechud)
+				return istype(G, /obj/item/clothing/glasses/hud/security) && G.active || istype(G, /obj/item/clothing/glasses/sunglasses/sechud) && G.active || istype(G, /obj/item/clothing/glasses/eyepatch/hud/security) && G.active || eye_hud == "Security"
```
Comment (NonQueueingMatt) [inline code/modules/mob/living/carbon/human/examine.dm:375]: a better way to do this imho would honestly be to have a proc on glasses:
```
/obj/item/clothing/glasses/proc/is_security_hud()
	return FALSE
```
override this as needed on instances of glasses that count as security:
```
/obj/item/clothing/glasses/hud/security/is_security_hud()
	return TRUE && active
```
and so on for medical as well

**Comment by Sparkyhotdog** [comment]
Working with the procs, as suggested

**Review by skull132 — CHANGES_REQUESTED** [review]
Not a fan of the magic string shenanigans. It'll lead to typos and stuff causing code to not work.

Matt's idea of `is_medical_hud()` and `is_security_hud()` is better, as the procs are concrete and simply return booleans. Alternatively, use defines to hide the magic strings.

File: code/modules/clothing/glasses/glasses.dm [hunk 3]
```diff
@@ -76,6 +76,12 @@ BLIND     // can't see anything
 	user.update_inv_l_hand(0)
 	user.update_inv_r_hand(1)
 
+/obj/item/clothing/glasses/proc/is_sec_hud()
+	return 0
+
+/obj/item/clothing/glasses/proc/is_med_hud()
+	return
```
Comment (skull132) [inline code/modules/clothing/glasses/glasses.dm:83]:
```suggestion
	return FALSE
```

File: code/modules/clothing/glasses/glasses.dm [hunk 4]
```diff
@@ -76,6 +76,12 @@ BLIND     // can't see anything
 	user.update_inv_l_hand(0)
 	user.update_inv_r_hand(1)
 
+/obj/item/clothing/glasses/proc/is_sec_hud()
+	return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:80]:
```suggestion
	return FALSE
```

File: code/modules/clothing/glasses/glasses.dm [hunk 5]
```diff
@@ -76,6 +76,12 @@ BLIND     // can't see anything
...
+
+/obj/item/clothing/glasses/proc/is_med_hud()
+	return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:83]:
```suggestion
	return FALSE
```

File: code/modules/clothing/glasses/glasses.dm [hunk 6]
```diff
@@ -471,6 +477,12 @@ BLIND     // can't see anything
 	src.hud = new/obj/item/clothing/glasses/hud/security(src)
 	return
 
+/obj/item/clothing/glasses/sunglasses/sechud/is_sec_hud()
+	if (active)
+		return 1
+	else
+		return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:484]:
```suggestion
	return active
```

File: code/modules/clothing/glasses/glasses.dm [hunk 7]
```diff
@@ -514,10 +526,12 @@ BLIND     // can't see anything
 		on = !on
 		if(on)
 			flash_protection = FLASH_PROTECTION_NONE
+			active = 1
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:529]:
```suggestion
			active = TRUE
```

File: code/modules/clothing/glasses/glasses.dm [hunk 8]
```diff
@@ -514,10 +526,12 @@ BLIND     // can't see anything
...
 			src.hud = hud_holder
 			to_chat(user, "You switch \the [src] to HUD mode.")
 		else
 			flash_protection = initial(flash_protection)
+			active = 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:534]:
```suggestion
			active = FALSE
```

File: code/modules/clothing/glasses/glasses.dm [hunk 9]
```diff
@@ -679,6 +693,12 @@ BLIND     // can't see anything
 /obj/item/clothing/glasses/eyepatch/hud/security/process_hud(var/mob/M)
 	process_sec_hud(M, 1)
 
+/obj/item/clothing/glasses/eyepatch/hud/security/is_sec_hud()
+	if (active)
+		return 1
+	else
+		return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:700]:
```suggestion
	return active
```

File: code/modules/clothing/glasses/glasses.dm [hunk 10]
```diff
@@ -688,6 +708,12 @@ BLIND     // can't see anything
 /obj/item/clothing/glasses/eyepatch/hud/medical/process_hud(var/mob/M)
 	process_med_hud(M, 1)
 
+/obj/item/clothing/glasses/eyepatch/hud/medical/is_med_hud()
+	if (active)
+		return 1
+	else
+		return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/glasses.dm:715]:
```suggestion
	return active
```

File: code/modules/clothing/glasses/hud.dm [hunk 11]
```diff
@@ -23,6 +23,12 @@
 /obj/item/clothing/glasses/hud/health/process_hud(var/mob/M)
 	process_med_hud(M, 1)
 
+/obj/item/clothing/glasses/hud/health/is_med_hud()
+	if (active)
+		return 1
+	else
+		return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/hud.dm:30]:
```suggestion
	return active
```

File: code/modules/clothing/glasses/hud.dm [hunk 12]
```diff
@@ -49,6 +55,12 @@
 	body_parts_covered = 0
 	var/global/list/jobs[0]
 
+/obj/item/clothing/glasses/hud/security/is_sec_hud()
+	if (active)
+		return 1
+	else
+		return 0
```
Comment (mikomyazaki) [inline code/modules/clothing/glasses/hud.dm:62]:
```suggestion
	return active
```

File: code/modules/mob/living/carbon/human/examine.dm [hunk 13]
```diff
@@ -364,13 +364,22 @@
 /proc/hasHUD(mob/M as mob, hudtype)
 	if(istype(M, /mob/living/carbon/human))
```
Comment (mikomyazaki) [inline code/modules/mob/living/carbon/human/examine.dm:365]:
```suggestion
/proc/hasHUD(mob/M, hudtype)
	if(ishuman(M))
```

File: code/modules/mob/living/carbon/human/examine.dm [hunk 14]
```diff
@@ -364,13 +364,22 @@
...
-				return 0
+		var/obj/item/clothing/glasses/G = H.glasses
+		var/eye_hud = 0
+		var/hud = 0
+		// Checks for eye sensor HUD
+		var/obj/item/organ/internal/augment/eye_sensors/E = locate() in H.internal_organs
+		if(E)
+			eye_hud = E.check_hud(hudtype)
+		if(G)
+			switch(hudtype)
```
Comment (mikomyazaki) [inline code/modules/mob/living/carbon/human/examine.dm:375]: Would be preferable if the strings "security" and "medical" were defines, if you want to do that that'd be cool.

File: code/modules/mob/living/carbon/human/examine.dm [hunk 15]
```diff
@@ -364,13 +364,22 @@
...
+				if("security")
+					hud = G.is_sec_hud()
+				if("medical")
+					hud = G.is_med_hud()
+				else
+					return 0
```
Comment (mikomyazaki) [inline code/modules/mob/living/carbon/human/examine.dm:381]:
```suggestion
```

Wearing hudless glasses will block hud-eye hud-vision if these lines are kept.

File: code/modules/organs/subtypes/augment.dm [hunk 16]
```diff
@@ -251,6 +251,12 @@
 		return
 	E.take_damage(5)
 
+/obj/item/organ/internal/augment/eye_sensors/proc/check_hud(var/hud)
+	if(hud == lowertext(selected_hud))
+		return 1
+	else
+		return 0
```
Comment (mikomyazaki) [inline code/modules/organs/subtypes/augment.dm:258]:
```suggestion
	return (hud == lowertext(selected_hud))
```

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/examine.dm (hunk 15 above)
Comment (Sparkyhotdog) [reply code/modules/mob/living/carbon/human/examine.dm:381]: That doesn't do that, it means that if the proc is called (which currently is either to check if security or medical HUDs are active) and the requested hud type wasn't security or medical, it stops, but its a redundant line anyway.

Comment (mikomyazaki) [reply code/modules/mob/living/carbon/human/examine.dm:381]: Ah, right. Thanks 👍

File: code/modules/organs/subtypes/augment.dm [hunk 17]
```diff
@@ -208,6 +208,7 @@
 	activable = TRUE
 	organ_tag = BP_AUG_EYE_SENSORS
 	action_button_name = "Toggle Eye Sensors"
+	var/active_hud = "disabled"
```
Comment (mikomyazaki) [inline code/modules/organs/subtypes/augment.dm:211]: I meant we can make hud_types be lowercase. We don't need this active_hud var.

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/modules/organs/subtypes/augment.dm (hunk 17 above)
Comment (Sparkyhotdog) [reply code/modules/organs/subtypes/augment.dm:211]: active_hud is a bug_fix, as if you select security without security access, that is still the selected_hud, meaning one gets sec records without the HUD.

File: code/_helpers/lists.dm [hunk 18]
```diff
@@ -733,3 +733,9 @@
 	. = list()
 	for(var/e in L)
 		. += L[e]
+
+/proc/capitalize_list(var/list/L)
+	. = list()
+	for (var/string in L)
+		. += capitalize(string)
+	return .
```
Comment (mikomyazaki) [inline code/_helpers/lists.dm:741]:
```suggestion
```
As discussed on Discord.

**Review by mikomyazaki — CHANGES_REQUESTED** [review]

File: code/modules/organs/subtypes/augment.dm (hunk 17 above)
Comment (mikomyazaki) [reply code/modules/organs/subtypes/augment.dm:211]: I guess that is one way to fix that. Okay.

**Review by mikomyazaki — APPROVED** [review]

**Pull request merged**